`stdlib_system`: essential path functionality (#999)
fortran-lang/stdlib: doc/specs/stdlib_system.md

```fortran
{!example/system/example_delete_file.f90!}
```

> ## `joinpath` - Joins the provided paths according to the OS

Reviewer: Quoting https://github.com/fortran-lang/stdlib/pull/939#pullrequestreview-2638507159, I agree it would be a good idea to keep the naming convention as homogeneous as possible. In this case it would be `join_path` with snake case.

### Status

Experimental

### Description

This interface joins the paths provided to it according to the platform specific path-separator.
i.e `\` for windows and `/` for others

### Syntax

`res = ` [[stdlib_system(module):join_path(interface)]] ` (p1, p2)`

`res = ` [[stdlib_system(module):join_path(interface)]] ` (p)`

### Class
Pure function

### Arguments

> `p1, p2`: Shall be a character string. It is an `intent(in)` argument.

Reviewer: Because stdlib offers a string type, it would be nice if these string manipulation functions were `interfaces` working with either character strings (as currently proposed), or the internal `string_type` type.

Author: I made these functions interfaces keeping this possibility in mind... But what is considered a good way of doing this? Writing procedures for `string_type` and then, when calling them with `character` variables just assign these to the new `string_type` variables and call the `string_type` version of the interface itself or code duplication, keeping the code separate for these two although they are in essence the same?

Author: Ah, I got it!

Reviewer: Yes: the easiest option is to keep the `character` implementation and then return the result to a string via `move`: https://github.com/fortran-lang/stdlib/blob/d8a90aa878afb4cc315d81384d9077285a6ebd2d/src/stdlib_string_type.fypp#L220-L227

Reviewer: It works @wassup05, however, there are two copies: `string->char` for the input arguments, and then `char -> string` on the assignment. Using `move` would at least avoid incurring the last copy.

Reviewer: Because the internal `string_type` representation is just an allocatable character variable, I would also support (to be discussed with the community as a separate PR) the implementation of a tiny zero-copy "view" function for the string type, [see here](https://godbolt.org/z/dvaK7Tc1E): 

```fortran
    !> Zero-copy view of the string as a character array
    pure function view(string) result(char_array)
        type(string_type), intent(in), target :: string
        character(:), pointer :: char_array
        if (allocated(string%raw)) then
            char_array => string%raw
        else
            nullify(char_array)
        end if
    end function view
```

    or
`p`: Shall be a list of character strings. `intent(in)` argument.

### Return values

The resultant path.

## `operator(/)`

Alternative syntax to`join_path` using an overloaded operator. Join two paths according to the platform specific path-separator.

### Status

Experimental

### Syntax

`p = lval / rval`

### Class

Pure function.

### Arguments

`lval`: A character string, `intent(in)`.

`rval`: A character string, `intent(in)`.

### Result value

The result is an `allocatable` character string

#### Example

```fortran
{!example/system/example_path_join.f90!}
```

## `split_path` - splits a path immediately following the last separator

### Status

Experimental

### Description

This subroutine splits a path immediately following the last separator after removing the trailing separators
splitting it into most of the times a directory and a file name.

### Syntax

`call `[[stdlib_system(module):split_path(interface)]]`(p, head, tail)`

### Class
Subroutine

### Arguments

> `p`: A character string containing the path to be split. `intent(in)`
> `head`: The first part of the path. `allocatable, intent(out)`
> `tail`: The rest part of the path. `allocatable, intent(out)`

Reviewer: ```suggestion
`p`: A character string containing the path to be split. It is an `intent(in)` argument.
`head`: The first part of the path. It is an `allocatable, intent(out)` argument.
`tail`: The rest part of the path. It is an `allocatable, intent(out)` argument.
```

### Behavior

> - If `p` is empty, `head` is set to `.` and `tail` is empty

Reviewer: ```suggestion
- If `p` is empty, `head` is set to `.` and `tail` is left empty.
```

> - If `p` consists entirely of path-separators. `head` is set to the path-separator and `tail` is empty
> - `head` ends in a path-separator if and only if `p` appears to be a root directory or child of one

Reviewer: ```suggestion
- If `p` consists entirely of path-separators, `head` is set to the path-separator and `tail` is left empty.
- `head` ends with a path-separator if and only if `p` appears to be a root directory or child of one.
```

### Return values

The splitted path. `head` and `tail`.

### Example

```fortran
{!example/system/example_path_split_path.f90!}
```

## `base_name` - The last part of a path

### Status

Experimental

### Description

This function returns the last part of a path after removing trailing path separators.

### Syntax

`res = ` [[stdlib_system(module):base_name(interface)]]`(p)`

### Class
Function

### Arguments

> `p`: the path, a character string, `intent(in)`

Reviewer: ```suggestion
`p`: the path, a character string. It is an `intent(in)` argument.
```

### Behavior

- The `tail` of `stdlib_system(module):split_path(interface)` is exactly what is returned. Same Behavior.

### Return values

A character string.

### Example

```fortran
{!example/system/example_path_base_name.f90!}
```

## `dir_name` - Everything except the last part of the path

### Status

Experimental

### Description

This function returns everything except the last part of a path.

### Syntax

`res = ` [[stdlib_system(module):dir_name(interface)]]`(p)`

### Class
Function

### Arguments

> `p`: the path, a character string, `intent(in)`

Reviewer: ```suggestion
`p`: the path, a character string. It is an `intent(in)` argument.
```

### Behavior

> - The `head` of `stdlib_system(module):split_path(interface)` is exactly what is returned. Same Behavior.

Reviewer: please add a link to `split_path` in FORD syntax.
